SHARE-122 readded & fixed codeview object image

dmetzner opened the pull request:
on every project page when looking at the code view there should
now be the correct object image displayed for each object.
Also added a fontawesome icon (times) when there is no image found.

dmetzner committed:
SHARE-122 readded & fixed codeview object image

on every project page when looking at the code view there should
now be the correct object image displayed for each object.
Also added a fontawesome icon (times) when there is no image found.

outsourced the codeview js code

dmetzner commented:
![image](https://user-images.githubusercontent.com/40868718/63496775-c6d41c80-c4c2-11e9-9470-bd7445b21939.png)

mfuerni commented on src/Catrobat/Services/CatrobatCodeParser/ParsedObjectAsset.php:
```diff
@@ -31,16 +33,17 @@ public function __construct(\SimpleXMLElement $asset_xml_properties)
 
-    $this->file_name = $asset_xml_properties->fileName;
+    $this->file_name = $asset_xml_properties['fileName'];
+    $this->file_name = str_replace('#', '%23', $this->file_name);
```
Shouldn't we use `urlencode()` here instead? Or better use twig `url_encode` filter instead.

mfuerni commented on templates/Program/codeview.html.twig:
```diff
@@ -6,7 +6,19 @@
+    {% if scene_name is null %}
+      <img src="{{ asset('./'~path~'images/'~object.getLooks()[0].getFileName()) }}" class="object-icon">
+    {% else %}
+      <img src="{{ asset('./'~path~'/'~scene_name~'/images/'~object.getLooks()[0].getFileName()) }}" class="object-icon">
```
`path` ends with `/`, so use `~path~scene_name~` instead

mfuerni commented on templates/Program/codeview.html.twig:
```diff
@@ -36,11 +49,27 @@
+            {% if scene_name is null %}
+              <img src="{{ asset('./'~path~'images/'~script.getLookFileName()) }}" class="brick-look script-brick-look">
+            {% else %}
+              <img src="{{ asset('./'~path~'/'~scene_name~'/images/'~script.getLookFileName()) }}"
```
`path` ends with `/`, so use `~path~scene_name~` instead

mfuerni commented on templates/Program/codeview.html.twig:
```diff
@@ -36,11 +49,27 @@
+                {% if scene_name is null %}
+                  <img src="{{ asset('./'~path~'images/'~brick.getLookFileName()) }}" class="brick-look">
+                {% else %}
+                  <img src="{{ asset('./'~path~'/'~scene_name~'/images/'~brick.getLookFileName()) }}"
```
`path` ends with `/`, so use `~path~scene_name~` instead

mfuerni commented on src/Catrobat/Services/CatrobatCodeParser/ParsedObjectAsset.php:
```diff
@@ -31,16 +33,17 @@ public function __construct(\SimpleXMLElement $asset_xml_properties)
     $this->asset_xml_properties = $asset_xml_properties;
 
-    if (count($asset_xml_properties->name) !== 0)
+    if (count($this->file_name = $asset_xml_properties['name']) !== 0)
```
I don't get the idea of these two changes (if condition + body).

I think the following code is enough for this method, since name and fileName are always attributes, not nodes.
```
    $this->asset_xml_properties = $asset_xml_properties;
    $this->name = $asset_xml_properties[Constants::NAME_ATTRIBUTE];
    $this->file_name = urlencode($asset_xml_properties['fileName']);
```

dmetzner replied:
I have tested it; it is enough. But I have used rawurlencode() which uses '%20' instead of '+' for whitespaces

dmetzner replied:
just had to change the outdated code.xml and unit test.

mfuerni reviewed (changes requested):
The aspect ratio of the images is not kept. Maybe we should change that.
Do we need any tests for the presence of the correct image?

✓ All tests are passed

dmetzner commented:
changed css to 'object-fit: cover' to keep the aspect ratio.
added a twig extension to check if the asset exists. If the correct images doesn't exist the object-not-found placeholder is shown.

dmetzner commented:
- added checks for scripts, bricks and sounds, 
- fixed brick, scripts, sound getFileName
- outsourced codeview js

mfuerni commented on src/Catrobat/Services/CatrobatCodeParser/Scripts/WhenBGChangeScript.php:
```diff
@@ -25,14 +25,14 @@ protected function create()
       else
       {
-        $this->look_file_name = $this->script_xml_properties->look->fileName;
+        $this->look_file_name = $this->brick_xml_properties->look['fileName'];
```
`brick_xml_properties` does not exist in class `Script`. Must be changed to `script_xml_properties`.

mfuerni commented on templates/Program/codeview.html.twig:
```diff
@@ -37,19 +58,37 @@
             <div class="brick" style="background-image: url({{ asset('images/bricks/'~brick.getImgFile()) }})">
               <div class="brick-text">{{ brick.getCaption() }}</div>
+              {% if brick.getLookFileName() is defined %}
+                <img src="{{ asset('./'~path~scene_name~'/images/'~brick.getLookFileName()) }}"
```
I think this line (75) should be removed.

mfuerni reviewed (changes requested):
Just two minor fixes, then I can approve it.
